### ISSUE TITLE
the importance of having a coordinate system independent of what the source files offer

### ISSUE THREAD

**Issue by proycon** [annotation/stam#9]
I was thinking about \"the importance of having a coordinate system independent
of what the source files offer\" which @dirkroorda mentioned the other day, and
which was also described in the [Unlocking Digital Texts Position
Paper](https://osf.io/u6vb4):

> From these formats, it is definitely possible to introduce a “glyph1-level” fragment addressing
> scheme, comprising an offset from the start of the file. This effectively reduces all text
> formats to plain-text by stripping away any additional tagging and non-textual components.
>
> This is not an entirely trivial exercise, since some additional complexities around Unicode
> normalisation rules and white-space handling will need to be dealt with, in order to ensure
> that plain-text conversions are carried out in a consistent manner.
>
> However, at this stage, it appears that it would be advantageous to also have a
> higher level scheme that operates in a more “human-friendly” way, with word (or token)
> granularity and some sense of semantic structure at a level similar to Markdown or a
> light-TEI schema. 

How does this relate to STAM? We have higher-order annotations that allows modelling higher-level schemes. 
We can annotate a sentence and then annotate a word in that sentence using relative offsets:

![explicit containment](https://raw.githubusercontent.com/annotation/stam/master/examples/explicit_containment_overview.png)

The offsets still refer to the unicodepoint level, but no longer relative to
the resource as a whole but to the annotation that is being pointed at (the
sentence in this case).

The recent proposal for the [STAM Baseoffset
extension](https://github.com/annotation/stam/pull/7) is also relevant in this
because it allows us to use a start/base offset that deviates from the actual
text (a simple decoupling from the actual coordinate system, though the units
are still the same).

Next we also have our `CompositeSelector` (and `MultiSelector`) that would let us
model things the other way round, we can have the sentence be the higher-order
annotation and have it point annotations that are words, and those in turn point to the resource
(using offsets).

At this point a question arises of something we can't model in STAM yet. Our
offsets are always unicode points (as that's our most atomic unit). If you want
to address things at a higher level like described in the previous paragraph
then that requires explicitly enumerating all the targets in a
CompositeSelector/MultiSelector/DirectionalSelector. But what if we want to use
offsets in another coordinate system here? Say a selector that selects the
second up to the ninth *word*? Do we want a selector that can express this
*whilst automatically interpolating the points in between*?

Adding something like that should be possible and adds more flexibility to how
people can use STAM for modelling, but it comes at the cost of adding further
complexity to STAM. So probably it should be an extension. 

Eventually we could even go as far as have a universal Selector that points to
something (resource/annotation) that is the result of a whole query. That might
subsume the above use-case as well, but would rely on several extensions (most
notably the query system which will be upcoming anyway) that are not trivial.

Text-Fabric and FoLiA both rely in the core on a coordinate system
more detached from the text, in both a text is merely an annotation like anything else. 

The situation in STAM is a bit different, almost everything is an annotation
but the text itself is the primary thing an annotation points to (a slice
thereof), either directly or indirectly. I do think that's the proper method
for a standoff text annotation model.

Last, a word about complex selectors like those in the Web Annotation model
which can reference XPath and other complex file formats. I do consider these
explicitly out of scope for STAM. We want to untangle text and annotations
completely, so text is its most bare form (plain text, utf-8) and all
annotations reference that, rather than some hybrid.

I just wanted to throw all this out here to voice and hear some thoughts and if
needed have some discussion, I'm especially interested in what @dirkroorda
thinks.

**Comment by dirkroorda** [comment]
@proycon ,

I agree with you that the choice to use annotations of annotations in STAM is the right way forward.

If you select words in a sentence, then you make an annotation that either

1. targets the words as annotations (so 7 items are selected, each of which is a word annotation)

or

2. targets the text that is targeted by the targeted words (if there is intervening whitespace and punctuation between the words, that is not targeted)

or

3. targets the text from the start of the text targeted by the first targeted word till the end of the text targeted by the last targeted word, including intervening space.

I think 2 and 3 are wrong.

In interpretation 3, if you target word 2 and word 8, but not the intervening words it still would target
the whole stretch from word 2 till word 8 (including), and that is odd.

Interpretation 2 looks a bit like 1 and for words it does not matter that much, but I think if you target words you also want to access other annotations on words, such as part-of-speech.

What is worse, it does not generalise neatly to other things than words. Suppose sentences are built from phrases, but phrases can be nested, and can have gaps. Like in:

```
Ik denk niet dat de man thuis kan komen die ik verwacht.
11 2222 333333333333333333333333333333333333333333333333
        4444 555 666666 777777777777777 6666666666666666
                        88888 999 88888 aaa bb ccccccccc
```

This is only a possible analysis of the sentence. Note that `thuiskomen` is one word, that `niet` occurs in the main clause but negates the verb in the subordinate clause, and that 
`de man die ik verwacht` is intruded by the predicate `thuis kan komen`.

We have 12 phrases here. What does it mean to target phrases 3-9 in this interpretation?
That is odd, because phrase 9 is contained in phrase 3. Under interpretation 3, 
literally applied, phrases 3-8 select more text than phrases 3-9 do !

Ok, but at some point we do need to compute where the selected annotations land in the text, and how to display them.

And there you have a lot of choices. In Text-Fabric I have developed a way to do exactly that, and a bit more, and it is not trivial, but it is not the responsibility of the model
to define this.

Suppose you want to display a sentence annotation, and there are also clause, phrase, word annotations that land in the same piece of text. Then you could highlight the relevant text in different colours and display the boundaries of those annotations as well, and possibly some annotation bodies too. You could make controls on the interface, so that the user can select what level of detail should be displayed.

Summing up: why not stick to interpretation 1: annotations of annotations have selectors that target annotations, and do not go directly to the text.

Computing the exact text that falls under the targeted annotations is not task of the model, but can be done by a [display algorithm](https://annotation.github.io/text-fabric/tf/about/displaydesign.html). 

Likewise, a *query language* is responsible for having primitives to look for annotations whose textual landing places have certain
[spatial relationships](https://annotation.github.io/text-fabric/tf/about/searchusage.html#relational-operators).

**Comment by dirkroorda** [comment]
There is another approach, one that I have adopted in Text-Fabric.

Let all segment annotations target the text directly.

The disadvantage is that there is no quick/easy method to find the sub segments of segments. Likewise it is not easy to find the supersegments of segments.

Yet these are operations that should be quick and easy, because they are a low-level part of many operations, including displaying text-fragments.

So we have to *make* those operations easy.

In Text-Fabric I have the locality operations, by which you can find the nodes above, below, before and after a node. For example:

`L.d(n, otype="line")` finds all the lines whose targeted text is contained in the targeted text of `n`. This works not only for segmentation nodes but for *all* nodes.

See [locality](https://annotation.github.io/text-fabric/tf/core/locality.html#tf.core.locality.Locality)

In STAM you could do this by generating this as new, derived annotations.
That will be quite a bit of extra annotations.

In Text-Fabric, `L.d()` is a function, it does a bit of computing itself.
But without help, it would be inefficient, especially if you need to call `L.d()` many times in the same context.

That is why I help `L.d()` by generating data and storing it next to the optimized tf data. See [here](https://annotation.github.io/text-fabric/tf/core/computed.html#precomputed-data-components).

The data for `L.d()` is prepared here: [prepare](https://annotation.github.io/text-fabric/tf/core/prepare.html#tf.core.prepare.levDown).

I think any performant application (based on stand-off annotations) that provides rich displays of arbitrary corpus fragments, needs to find a solution for efficiently navigating the neighbourhood.

This is one of the things where XML is almost trivial.

**Comment by proycon** [comment]
Thanks for the elaborate response and analysis!

> I think 2 and 3 are wrong.

I agree that 1 is the best way to model things, also in line with what you said in the chat:

> De segmentatie annotaties moeten levels hebben die uit elkaar opbouwbaar zijn. Dat is een concept van Crist-Jan Doedens in zijn proefschrift: Text Databases - One Database Model and Several Retrieval Languages, verdedigd in 1994.

The example in the schema I showed would correspond to approach 1, I called it "explicit containment".
Closest to approach 2 would be what I call "implicit containment", shown in the following schema:

![implicit containment](https://raw.githubusercontent.com/annotation/stam/master/examples/implicit_containment_overview.png)

That's probably closer to the Text-Fabric approach.

As we try to offer a fair amount of flexibility in STAM, so I do think
this should still be possible to model if the user desires (as it is now). But perhaps you're
right that we don't need to spend too much effort to implement approach 3 if we consider
it a wrong approach. 

Your point about the intervening space is also a good one, if you target multiple words with a CompositeSelector in either approach 1 or 2, then the intervening space is not targeted indeed.

> Summing up: why not stick to interpretation 1: annotations of annotations have selectors that target annotations, and do not go directly to the text.

Yes, in principle I agree, although I don't want to force this on the modeller.

> We have 12 phrases here. What does it mean to target phrases 3-9 in this interpretation?
> That is odd, because phrase 9 is contained in phrase 3. Under interpretation 3, 
> literally applied, phrases 3-8 select more text than phrases 3-9 do !

Agreed, this indeeds adds a whole extra layer of complexity where we probably don't want to go. Let's scrap the idea of implementing approach 3.


> There is another approach, one that I have adopted in Text-Fabric.
>
> Let all segment annotations target the text directly.
>
> The disadvantage is that there is no quick/easy method to find the sub
> segments of segments. Likewise it is not easy to find the supersegments of
> segments.
>
> Yet these are operations that should be quick and easy, because they are a
> low-level part of many operations, including displaying text-fragments.
>
> So we have to *make* those operations easy.

I do have the same intention in the STAM library implementation; facilitating all these computations (using reverse indices etc) so you can quickly resolve any that resembles implicit containment (or other text relations).
We need such functionality even if we favour approach 1, because there will always be implicit relations to infer that are not explicitly modelled.

> Computing the exact text that falls under the targeted annotations is not
> task of the model, but can be done by a [display
> algorithm](https://annotation.github.io/text-fabric/tf/about/displaydesign.html). 

Perhaps not of the model per se, but definitely of the library; it should be
able to return the exact slice of the text any annotation references. Of course
any further display is up to higher-level function. You've spent quite some
time in Text Fabric on that, for the future I'd like to see that we can reuse
all that with a STAM-backend.

> Likewise, a *query language* is responsible for having primitives to look for annotations whose textual landing places have certain
> [spatial relationships](https://annotation.github.io/text-fabric/tf/about/searchusage.html#relational-operators).

Agreed, formulating these primitives will be part of a STAM extension (which I've thinking about a lot past week).
